compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs

```diff
@@ -417,7 +409,7 @@ impl<'tcx> ConstToPat<'tcx> {
                                 IndirectStructuralMatch { non_sm_ty: *pointee_ty },
                             );
                         }
-                        PatKind::Constant { value: cv }
+                        return Err(FallbackToConstRef);
```

Comment:
This is a change in behaviour. If we have

```rust
struct Inner;
impl PartialEq for Inner {
    fn eq(&self, _: &Self) -> bool {
        true
    }
}
impl Eq for Inner {}

#[derive(PartialEq, Eq)]
enum Eek<'a> {
    Foo,
    Bar(&'a Inner),
}

const THE_CONST: &Eek<'static> = &Eek::Bar(&Inner);
```

Previously `THE_CONST` was destructured to `Deref(Variant(Bar, CONST))` for an autogenerated `const CONST: &Inner = &Inner;`

Now it doesn't get destructured, but we directly see `THE_CONST`.

A const of value `&Eek::Foo` would still get destructured to `Deref(Variant(Foo, []))`

If we had something like

```rust
struct Inner;
impl PartialEq for Inner {
    fn eq(&self, _: &Self) -> bool {
        true
    }
}
impl Eq for Inner {}

#[derive(PartialEq, Eq)]
enum Eek<'a> {
    Foo,
    Bar(&'a Option<Inner>),
}

const THE_CONST: &Eek<'static> = &Eek::Bar(&None);
```

We would get `Deref(Variant(Bar, Deref(Variant(None, []))`, before and after this PR.

If you had a `Some` in there, we'd again use `THE_CONST` directly.

Reply:
> A const of value &Eek::Foo would still get destructured to Deref(Variant(Foo, []))

Wait, so the analysis here is value-driven, not type-driven?

Reply:
Yes, I believe requiring recursive structural equality on the type of constants in order to expand them into patterns is not doable without breaking real code

Reply:
Yeah, I can't say I am surprised.

This needs careful documentation though, to explain how this value analysis works.